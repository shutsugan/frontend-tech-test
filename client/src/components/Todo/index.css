.todo {}

.todo__info {
    color: var(--sub-title);
    font-size: .9rem;
}

.todo__remove-button {
    width: 16px;
    height: 16px;
    border-radius: 16px;
    background: var(--main);
    color: #fff;
    font-size: .7rem;
    font-weight: 600;
    cursor: pointer;
    transition: transform .15s ease-in;
}

.todo__remove-button:hover {
    transform: scale(1.1);
}

