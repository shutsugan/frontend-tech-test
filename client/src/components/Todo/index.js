import React from 'react';
import { useStoreActions } from 'easy-peasy';

import './index.scss';

const Todo = ({ id, title, description }) => {
    const selectTodo = useStoreActions(actions => actions.todos.selectTodo);
    const removeTodo = useStoreActions(actions => actions.todos.removeTodo);

    return (
        <div className="todo flex space-between pdt-8 br-bottom">
            <div>
            <h3 className="mr-none mrs-8">{ title }</h3>
            <p className="todo__info mr-none mrs-8">{ description }</p>
            </div>
            <div className="flex center pd-8">
                <span
                    className="link mrr-16"
                    onClick={_ => selectTodo({ id, title, description})}>
                    Edit
                </span>
                <div
                    className="todo__remove-button flex center"
                    onClick={_ => removeTodo(id)}>
                    x
                </div>
            </div>
        </div>
    );
};

export default Todo;