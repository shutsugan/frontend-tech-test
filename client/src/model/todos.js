import { action, computed, thunk } from "easy-peasy";

const base_url = 'http://localhost:9001';

const noop = _ => {};
export const save = async (axios, actions, method, url, callback = noop) => {
    const res = await axios[method](`${base_url}/task/${url}`);
    const { task, message } = await res.data;

    callback(task);
    actions.setMessage(message);
};

export const checkProps = (actions, title, description) => {
    if (!title && !description) return actions.setMessage('Fields are reuqired');
};

export default {
    todos: [],
    selectedTodo: null,
    message: null,

    list: computed(state => Object.values(state.todos)),

    //actions
    setMessage: action((state, message) => state.message = message),
    selectTodo: action((state, todo) => state.selectedTodo = todo),
    added: action((state, todo) => state.todos[todo.id] = todo),

    fetched: action((state, payload = []) => {
        state.todos = payload.reduce((acc, todo) => {
            acc[todo.id] = todo;

            return acc;
        }, {});
    }),

    //thunks
    fetchTodos: thunk(async (actions, _, { injections }) => {
        const { axios } = injections
        const res = await axios.get(`${base_url}/tasks`);
        const { tasks } = await res.data;

        actions.fetched(tasks);
    }),

    addTodo: thunk(async (actions, { title, description }, { injections }) => {
        const { axios } = injections

        checkProps(actions, title, description);
        save(
            axios,
            actions,
            'post',
            `create/${title}/${description}`,
            todo => actions.added(todo)
        );
    }),

    updateTodo: thunk(async (actions, { id, title, description }, { injections }) => {
        const { axios } = injections;

        checkProps(actions, title, description);
        save(
            axios,
            actions,
            'put',
            `update/${id}/${title}/${description}`,
            _ => {
                actions.fetchTodos();
                actions.selectTodo(null);
            }
        );
    }),

    removeTodo: thunk(async (actions, id, { injections }) => {
        const { axios } = injections;

        save(
            axios,
            actions,
            'delete',
            `delete/${id}`,
            _ => actions.fetchTodos()
        );
    })
};