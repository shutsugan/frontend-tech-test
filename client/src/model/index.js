import todos from './todos';
import { thunk } from 'easy-peasy';

export default {
    todos,
    initialiseTodos: thunk(async (actions, payload, { dispatch }) => {
        await dispatch.todos.fetchTodos();
    })
};