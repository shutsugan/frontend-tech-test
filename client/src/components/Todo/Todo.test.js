import React from 'react';
import renderer from 'react-test-renderer';
import { createStore, StoreProvider } from 'easy-peasy';
import { expect as chaiExpect } from 'chai';

import model from '../../model';
import Todo from './index';

const store = createStore(model);

describe('Todo component', () => {
    it ('should render todo componenet', () => {
        const tree = renderer.create(
            <StoreProvider store={store}>
                <Todo />
            </StoreProvider>
        ).toJSON();
        
        expect(tree).toMatchSnapshot();
        chaiExpect(tree).to.be.an('object');
        chaiExpect(tree).to.have.own.property('props');
        chaiExpect(tree.children.length).to.equal(2);
    });
});