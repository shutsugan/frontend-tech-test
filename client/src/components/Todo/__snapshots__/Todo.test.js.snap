// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Todo component should render todo componenet 1`] = `
<div
  className="todo flex space-between pdt-8 br-bottom"
>
  <div>
    <h3
      className="mr-none mrs-8"
    />
    <p
      className="todo__info mr-none mrs-8"
    />
  </div>
  <div
    className="flex center pd-8"
  >
    <span
      className="link mrr-16"
      onClick={[Function]}
    >
      Edit
    </span>
    <div
      className="todo__remove-button flex center"
      onClick={[Function]}
    >
      x
    </div>
  </div>
</div>
`;
